vm-allocator/src/system.rs

```diff
@@ -81,6 +84,16 @@ impl SystemAllocator {
         }
     }
 
+    /// Reserves the next available system device instance id number.
+    pub fn allocate_instance_id(&mut self) -> Option<u32> {
+        if let Some(id) = self.next_instance_id.checked_add(1) {
+            self.next_instance_id = id;
+            Some(id - 1)
+        } else {
+            None
+        }
+    }
+
```

Comment:
Ok, so this is exactly the same code as the irq one. I think instead we need to build a generic `IdAllocator` structure and implement `allocate` and `free` methods for it. `allocate` could simply do the `checked_add()` that you do here but eventually we could decide to do something smarter (and use freed IDs) but that would take more complex data structures.

Both the above system irq and id allocation would call into the `IdAllocator.allocate()`.

Reply:
Also, we need to make the system allocator thread safe. I think we should have one mutex per system resource, keep the address and id allocators lock less and take the locks at the `system.rs` level.

Reply:
@sameo 
@sameo 
> Also, we need to make the system allocator thread safe. I think we should have one mutex per system resource, keep the address and id allocators lock less and take the locks at the `system.rs` level.

Which codes will use system allocator except vm-device? Once device_manager wrap it with Arc Mutex, it can protect it?

Reply:
> Ok, so this is exactly the same code as the irq one. I think instead we need to build a generic `IdAllocator` structure and implement `allocate` and `free` methods for it. `allocate` could simply do the `checked_add()` that you do here but eventually we could decide to do something smarter (and use freed IDs) but that would take more complex data structures.
> 
> Both the above system irq and id allocation would call into the `IdAllocator.allocate()`.

I agree with this.  There needs free for those resource.

Reply:
> Which codes will use system allocator except vm-device? 

There may be other parts of a VMM requesting for e.g. some address space range (e.g. VFIO).

> Once device_manager wrap it with Arc Mutex, it can protect it?

You want the top level API to be Mutex protected, so that multi threaded callers don't have to all implement a Mutex wrapper on top of it. In most cases callers will want thread safe code, but if they really need lock less API, they can use the lower level ones (`AddressAllocator`, etc).

Reply:
> I agree with this. There needs free for those resource.

Indeed.

Reply:
I am not sure we can use this code as is with arc and mutex on each resource. Maybe we should provide a trait here instead? Allocating these resources is on the critical path (i.e. it adds to the boot time) for Firecracker.

Reply:
> I am not sure we can use this code as is with arc and mutex on each resource. Maybe we should provide a trait here instead? Allocating these resources is on the critical path (i.e. it adds to the boot time) for Firecracker.

@andreeaflorescu  
Sorry I didn't quite get the comments. Which part of codes does your comments for?

Reply:
The comment refers to the previous comment from @sameo. It is about having AddressAllocator and IdAllocator in a Mutex.